op-batcher/batcher/driver_test.go

```diff
@@ -249,20 +251,67 @@ func TestBatchSubmitter_ThrottlingEndpoints(t *testing.T) {
 			// Start throttling loop in a goroutine
 			go bs.throttlingLoop(&wg, pendingBytesUpdated)
 
-			// Send test data to trigger throttling
-			pendingBytesUpdated <- 20000 // Over threshold, should trigger throttling
+			// Simulate block loading by sending perodically on pendingBytesUpdated
+			blockLoadingCtx, cancelBlockLoading := context.WithCancel(context.Background())
+			go func() {
+				// Simulate block loading
+				for range time.NewTicker(100 * time.Millisecond).C {
+					select {
+					case <-blockLoadingCtx.Done():
+						return
+					default:
+						// Simulate block loading
+						pendingBytesUpdated <- 20000 // the value doesn't actually matter for this test
+					}
+				}
+			}()
+
+			require.Eventually(t,
+				func() bool {
+					// Check that all endpoints were called
+					for i := range healthyCalls {
+						if healthyCalls[i] == 0 {
+							return false
+						}
+					}
+					for i := range unHealthyCalls {
+						if unHealthyCalls[i] == 0 {
+							return false
+						}
+					}
+					return true
+				}, time.Second*20, time.Millisecond*10, "All endpoints should have been called within 2s")
 
-			require.EventuallyWithT(t, func(collect *assert.CollectT) {
-				// Check that all endpoints were called
-				for i := range healthyCalls {
+			startTestServerAtAddr := func(addr string, handler http.HandlerFunc) *httptest.Server {
+				ln, err := net.Listen("tcp", addr)
+				require.NoError(t, err, "Failed to create new listener for test server")
 
-					require.Greater(t, healthyCalls[i], 0, "Healthy Server %d should have been called", i)
-				}
-				for i := range unHealthyCalls {
-					require.Greater(t, unHealthyCalls[i], 0, "Unhealthy Server %d should have been called", i)
+				s := &httptest.Server{
+					Listener: ln,
+					Config:   &http.Server{Handler: handler},
 				}
-			}, time.Millisecond*2000, time.Millisecond*10, "All endpoints should have been called within 2s")
+				s.Start()
+				return s
+			}
+
+			// Take one of the healthy servers down, wait 2s and restart. Check it is called again.
+			if len(healthyServers) > 0 {
+				restartedServerCalled := false
+
+				addr := healthyServers[0].Listener.Addr().String()
+				healthyServers[0].Close()
+				time.Sleep(time.Second * 2)
+				startTestServerAtAddr(addr, createHTTPHandler(t, func() { restartedServerCalled = true }, false))
+				defer healthyServers[0].Close()
+				t.Log("restarted server at", addr)
+
+				require.Eventually(t, func() bool {
+					return restartedServerCalled
+				}, time.Second*2, time.Millisecond*10, "Restarted server should have been called within 2s")
+			}
+
 			// Clean up
+			cancelBlockLoading()
 			close(pendingBytesUpdated)
 			cancel()
```

Comment:
Even though this is just a test, it would be more idiomatic (and panic proof) to `defer cancelBlockLoading()` right after creating the context. The same actually already happens for `cancel()`, so this call can be removed here.

Reply:
Removed the duplicated `cancel()` here [1b270c9](https://github.com/ethereum-optimism/optimism/pull/15603/commits/1b270c925f14856214cbc0082bbfe4273262d8d7).

Since we need to cancel block loading before closing the `pendingBytesUpdated` channel, I encapsulated the cleanup logic into a callback for `t.Cleanup()` [63c955c](https://github.com/ethereum-optimism/optimism/pull/15603/commits/63c955ce95156de0d66c0063b0b4e3cb6ba70cc3).